src/lib/uri/uri.c

```diff
@@ -42,8 +47,13 @@ uri_find_param(const struct uri *uri, const char *name)
 static void
 uri_param_add_value(struct uri_param *param, const char *value)
 {
-	size_t size = (param->value_count + 1) * sizeof(char *);
-	param->values = xrealloc(param->values, size);
+	assert(param->value_count <= param->values_capacity);
+	if (param->value_count == param->values_capacity) {
+		param->values_capacity = param->values_capacity == 0 ? 16 :
+					 param->values_capacity * 2;
+		size_t size = param->values_capacity * sizeof(char *);
+		param->values = xrealloc(param->values, size);
+	}
```

Comment:
We quite often use exponentially growing arrays. Even in this patch alone, there are two such places. Maybe, add a helper to `trivia/util.h` that would add an entry to such an array, relocating it if necessary? Or even a new data structure for dynamically growing arrays?

Reply:
I guess that's the burden of using C (as opposed to C++, where we could simply use `std::vector` everywhere).

I suppose we just open a refactoring ticket regarding this problem.

Reply:
I don't think that using `ibuf` would make this code any more efficient, because we have to allocate the final result with malloc anyway.

Reply:
I don't mean just place those arrays of pointers into an `ibuf` or a `region`. How about eliminating a 10^5 `strdup()` calls (on given test case)?

Reply:
Hmm, can we use [`grp_alloc`](https://github.com/tarantool/tarantool/blob/master/src/lib/salad/grp_alloc.h) here?

Reply:
> Hmm, can we use [`grp_alloc`](https://github.com/tarantool/tarantool/blob/master/src/lib/salad/grp_alloc.h) here?

Probably not.

> I don't mean just place those arrays of pointers into an `ibuf` or a `region`. How about eliminating a 10^5 `strdup()` calls (on given test case)?

I see your point. Those `strdup` look bad, and using a dynamic array here wouldn't help us eliminate them. Okay, I'm fine with leaving this code as is (with explicit realloc, that is). Maybe, we'll add a dynamic array some time later.

Reply:
I'm not sure that it'll be optimal to implement a two phare parsing. I guess that a one phase parsing with ahead memory allocation (for all those small pieces of data) would win. At least (in my head) it should be good for the parsing itself (as opposite to how we return it for Lua).

Maybe we can do something even more exciting: eliminate string copying and just reference parts of the original strings. Somethings like https://github.com/rapidlua/jstr, but without destruction of the original string (track both start and end of a chunk). Or with a single copy before the destructive parsing.

To be honest, I don't know, whether we can do something smart for returning values into Lua. Maybe create strings on demand? Don't know.

Reply:
(I didn't see the last your comment, when wrote mine one.)

Reply:
> Maybe we can do something even more exciting: eliminate string copying and just reference parts of the original strings.

The problem is that uri parameters can be added dynamically, after parsing the initial string (note that `uri_add_param` is a public method). This is used to pass extra uri parameters in a table.

Reply:
> The problem is that uri parameters can be added dynamically, after parsing the initial string (note that `uri_add_param` is a public method). This is used to pass extra uri parameters in a table.

Good point. I guess we still can speed up parsing of a complete URI, but dynamic adding parameters will not be so optimal.

Reply:
>  I guess we still can speed up parsing of a complete URI.

Maybe later. This patch is already a good step forward, because it fixes the issue at hand. As of eliminating `strdup`, let's postpone it until somebody complains. After all, one would hardly pass thousands of uri parameters in real life. That said, I think this PR is ready for merge.

Reply:
> I don't mean just place those arrays of pointers into an ibuf or a region. How about eliminating a 10^5 strdup() calls (on given test case)?

Yeah, I actually thought that this impacts performance the most. Turned out not to be in this specific case, but it's still a performance issue.

Could you please open a supplementary ticket?

Reply:
> This patch is already a good step forward, because it fixes the issue at hand.

Of course, this PR is about one specific problem and it fixes the problem.